File: backend/internal/adapter/postgres/repository.go
package postgres

import (
	"context"
	"errors"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/sqwdvrt/siteParserForFreelans/backend/internal/domain"
)

// JobRepository реализует port.JobRepository для PostgreSQL.
type JobRepository struct {
	pool *pgxpool.Pool
}

// NewJobRepository создаёт репозиторий с пулом соединений.
func NewJobRepository(pool *pgxpool.Pool) *JobRepository {
	return &JobRepository{pool: pool}
}

// Save сохраняет job в БД. При дубликате по URL обновляет raw_html,
// реактивирует status='active' и возвращает существующий id.
func (r *JobRepository) Save(ctx context.Context, job *domain.Job) (int64, error) {
	var id int64
	err := r.pool.QueryRow(ctx, `
		INSERT INTO jobs (source, url, external_id, title, description, budget, skills, posted_at, raw_html)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
		ON CONFLICT (url) DO UPDATE SET
			raw_html = EXCLUDED.raw_html,
			status = 'active',
			last_seen_at = NOW()
		RETURNING id
	`,
		job.Source,
		job.URL,
		nullIfEmpty(job.ExternalID),
		job.Title,
		nullIfEmpty(job.Description),
		nullIfEmpty(job.Budget),
		job.Skills,
		job.PostedAt,
		job.RawHTML,
	).Scan(&id)
	if err != nil {
		return 0, err
	}
	return id, nil
}

// GetByID возвращает job по id. nil при отсутствии.
func (r *JobRepository) GetByID(ctx context.Context, id int64) (*domain.Job, error) {
	var j domain.Job
	err := r.pool.QueryRow(ctx, `
		SELECT id, source, url, COALESCE(external_id, ''), title, COALESCE(description, ''),
			COALESCE(budget, ''), skills, posted_at, raw_html, created_at
		FROM jobs
		WHERE id = $1
		  AND status = 'active'
	`, id).Scan(&j.ID, &j.Source, &j.URL, &j.ExternalID, &j.Title, &j.Description,
		&j.Budget, &j.Skills, &j.PostedAt, &j.RawHTML, &j.CreatedAt)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, nil
		}
		return nil, err
	}
	return &j, nil
}

// GetByIDs возвращает map[id]*Job для всех найденных id. Один SQL-запрос.
func (r *JobRepository) GetByIDs(ctx context.Context, ids []int64) (map[int64]*domain.Job, error) {
	if len(ids) == 0 {
		return map[int64]*domain.Job{}, nil
	}
	rows, err := r.pool.Query(ctx, `
		SELECT id, source, url, COALESCE(external_id, ''), title, COALESCE(description, ''),
			COALESCE(budget, ''), skills, posted_at, raw_html, created_at
		FROM jobs
		WHERE id = ANY($1)
		  AND status = 'active'
	`, ids)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	result := make(map[int64]*domain.Job, len(ids))
	for rows.Next() {
		var j domain.Job
		if err := rows.Scan(&j.ID, &j.Source, &j.URL, &j.ExternalID, &j.Title, &j.Description,
			&j.Budget, &j.Skills, &j.PostedAt, &j.RawHTML, &j.CreatedAt); err != nil {
			return nil, err
		}
		result[j.ID] = &j
	}
	return result, rows.Err()
}

// TouchSeenAt обновляет last_seen_at = NOW() только для активных jobs по URL.
// Истёкшие вакансии намеренно игнорируются, чтобы избежать гонки реактивации.
func (r *JobRepository) TouchSeenAt(ctx context.Context, url string) error {
	_, err := r.pool.Exec(ctx,
		`UPDATE jobs SET last_seen_at = NOW() WHERE url = $1 AND status = 'active'`, url)
	return err
}

// ExpireStaleJobs помечает активные jobs как 'expired' если last_seen_at старше olderThanDays дней.
func (r *JobRepository) ExpireStaleJobs(ctx context.Context, olderThanDays int) ([]int64, error) {
	rows, err := r.pool.Query(ctx, `
		UPDATE jobs SET status = 'expired'
		WHERE status = 'active'
		  AND last_seen_at < NOW() - make_interval(days => $1)
		RETURNING id
	`, olderThanDays)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var ids []int64
	for rows.Next() {
		var id int64
		if err := rows.Scan(&id); err != nil {
			return nil, err
		}
		ids = append(ids, id)
	}
	return ids, rows.Err()
}

// ExpireByURL мгновенно помечает активный job как 'expired' по URL.
func (r *JobRepository) ExpireByURL(ctx context.Context, url string) ([]int64, error) {
	tx, err := r.pool.Begin(ctx)
	if err != nil {
		return nil, err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	rows, err := tx.Query(ctx, `
		UPDATE jobs
		SET status = 'expired'
		WHERE url = $1 AND status = 'active'
		RETURNING id
	`, url)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var ids []int64
	for rows.Next() {
		var id int64
		if err := rows.Scan(&id); err != nil {
			return nil, err
		}
		ids = append(ids, id)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	if len(ids) == 0 {
		return ids, tx.Commit(ctx)
	}
	if _, err := tx.Exec(ctx, `
		DELETE FROM notifications
		WHERE job_id = ANY($1) AND status = 'pending'
	`, ids); err != nil {
		return nil, err
	}
	if err := tx.Commit(ctx); err != nil {
		return nil, err
	}
	return ids, nil
}

// ExistsByURL проверяет наличие активного job по URL.
// Expired jobs не считаются существующими: при следующем обходе краулер
// повторно скачает детали и реактивирует job через Save().
func (r *JobRepository) ExistsByURL(ctx context.Context, url string) (bool, error) {
	var exists bool
	err := r.pool.QueryRow(ctx, `SELECT EXISTS(SELECT 1 FROM jobs WHERE url = $1 AND status = 'active')`, url).Scan(&exists)
	return exists, err
}

// GetUnembeddedIDs returns IDs of jobs without a job_embeddings entry (orphaned jobs).
func (r *JobRepository) GetUnembeddedIDs(ctx context.Context, limit int) ([]int64, error) {
	if limit <= 0 {
		limit = 100
	}
	rows, err := r.pool.Query(ctx, `
		SELECT j.id FROM jobs j
		LEFT JOIN job_embeddings je ON je.job_id = j.id
		WHERE je.job_id IS NULL
		  AND j.status = 'active'
		ORDER BY j.created_at DESC
		LIMIT $1
	`, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var ids []int64
	for rows.Next() {
		var id int64
		if err := rows.Scan(&id); err != nil {
			return nil, err
		}
		ids = append(ids, id)
	}
	return ids, rows.Err()
}

func nullIfEmpty(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

File: ai-service/src/ai_service/adapter/postgres/repository.py
"""PostgreSQL implementation of JobRepository."""

from __future__ import annotations

import json
from typing import Any

from pgvector import Vector
from psycopg2.extras import RealDictCursor

from ai_service.adapter.postgres._pooled_repository import PooledPostgresRepository
from ai_service.domain.job import Job
from ai_service.port.repository import JobEmbeddingRecord, JobRepository


class PostgresJobRepository(PooledPostgresRepository, JobRepository):
    """JobRepository через PostgreSQL."""

    def __init__(
        self,
        dsn: str,
        *,
        minconn: int = 1,
        maxconn: int = 10,
        statement_timeout_ms: int = 10_000,
    ) -> None:
        super().__init__(dsn, minconn=minconn, maxconn=maxconn, statement_timeout_ms=statement_timeout_ms)

    def get(self, job_id: int) -> Job | None:
        """Получить job по id."""
        with self._conn() as conn:
            with conn.cursor(cursor_factory=RealDictCursor) as cur:
                cur.execute(
                    """
                    SELECT
                        id, source, url, title, description, COALESCE(budget, '') AS budget,
                        raw_html, posted_at, created_at
                    FROM jobs
                    WHERE id = %s
                    """,
                    (job_id,),
                )
                row: dict[str, Any] | None = cur.fetchone()
        if row is None:
            return None
        return Job(
            id=row["id"],
            source=row["source"] or "kwork",
            url=row["url"] or "",
            title=row["title"] or "",
            description=row["description"],
            budget=row["budget"] or "",
            raw_html=row["raw_html"] or "",
            posted_at=row["posted_at"],
            created_at=row["created_at"],
        )

    def save_embedding(
        self,
        job_id: int,
        embedding: list[float],
        metadata: dict,
    ) -> None:
        """Сохранить embedding в job_embeddings."""
        with self._conn() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    """
                    INSERT INTO job_embeddings (job_id, embedding, ai_metadata)
                    VALUES (%s, %s, %s::jsonb)
                    ON CONFLICT (job_id) DO UPDATE SET
                        embedding = EXCLUDED.embedding,
                        ai_metadata = EXCLUDED.ai_metadata
                    """,
                    (job_id, embedding, json.dumps(metadata)),
                )

    def has_embedding(self, job_id: int) -> bool:
        """Проверить наличие embedding для job_id."""
        with self._conn() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    "SELECT 1 FROM job_embeddings WHERE job_id = %s LIMIT 1",
                    (job_id,),
                )
                return cur.fetchone() is not None

    def get_embedding(self, job_id: int) -> JobEmbeddingRecord | None:
        """Load saved embedding and ai_metadata for retry-safe reprocessing."""
        with self._conn() as conn:
            with conn.cursor(cursor_factory=RealDictCursor) as cur:
                cur.execute(
                    """
                    SELECT embedding, COALESCE(ai_metadata, '{}'::jsonb) AS ai_metadata
                    FROM job_embeddings
                    WHERE job_id = %s
                    LIMIT 1
                    """,
                    (job_id,),
                )
                row: dict[str, Any] | None = cur.fetchone()
        if row is None:
            return None
        metadata = row["ai_metadata"] if isinstance(row["ai_metadata"], dict) else {}
        raw_embedding = row["embedding"]
        return JobEmbeddingRecord(
            embedding=list(raw_embedding) if raw_embedding is not None else [],
            metadata=metadata,
        )

    def has_recent_similar_title(self, job_id: int, title: str, days: int = 7) -> bool:
        normalized_title = " ".join((title or "").strip().lower().split())
        if not normalized_title:
            return False
        with self._conn() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    """
                    SELECT EXISTS (
                        SELECT 1
                        FROM jobs j
                        WHERE j.id <> %s
                          AND j.created_at >= NOW() - make_interval(days => %s)
                          AND regexp_replace(
                                regexp_replace(lower(COALESCE(j.title, '')), '[^[:alnum:][:space:]]+', ' ', 'g'),
                                '\\s+',
                                ' ',
                                'g'
                              ) = regexp_replace(
                                regexp_replace(lower(%s), '[^[:alnum:][:space:]]+', ' ', 'g'),
                                '\\s+',
                                ' ',
                                'g'
                              )
                    )
                    """,
                    (job_id, days, normalized_title),
                )
                row = cur.fetchone()
        return bool(row[0]) if row else False

    def get_with_scores(
        self,
        job_ids: list[int],
        user_embedding: list[float],
        user_id: int | None = None,
    ) -> list[tuple[Job, float]]:
        """Load jobs by ids with similarity score against user embedding."""
        if not job_ids:
            return []
        if user_embedding is None or len(user_embedding) == 0 or len(user_embedding) != 384:
            return []

        with self._conn() as conn:
            with conn.cursor(cursor_factory=RealDictCursor) as cur:
                cur.execute(
                    """
                    SELECT
                        j.id,
                        j.source,
                        j.url,
                        j.title,
                        j.description,
                        COALESCE(j.budget, '') AS budget,
                        j.raw_html,
                        j.posted_at,
                        j.created_at,
                        ARRAY(
                            SELECT lower(trim(tag.value))
                            FROM jsonb_array_elements_text(
                                COALESCE(je.ai_metadata->'classification'->'technologies', '[]'::jsonb)
                            ) AS tag(value)
                            WHERE trim(tag.value) <> ''
                        ) AS technologies,
                        1 - (je.embedding <=> %s) AS similarity,
                        COALESCE(p.match_score, 1 - (je.embedding <=> %s)) AS match_score,
                        COALESCE(p.raw_similarity, 1 - (je.embedding <=> %s)) AS raw_similarity,
                        COALESCE(p.final_score, COALESCE(p.match_score, 1 - (je.embedding <=> %s))) AS final_score,
                        COALESCE(p.ranker_version, '') AS ranker_version,
                        COALESCE(p.reason_codes, ARRAY[]::text[]) AS reason_codes
                    FROM jobs j
                    JOIN job_embeddings je ON je.job_id = j.id
                    LEFT JOIN pending_ac_jobs p
                      ON p.job_id = j.id
                     AND (%s IS NULL OR p.user_id = %s)
                    WHERE j.id = ANY(%s)
                    ORDER BY COALESCE(p.final_score, p.match_score, 1 - (je.embedding <=> %s)) DESC, similarity DESC
                    """,
                    (
                        Vector(user_embedding),
                        Vector(user_embedding),
                        Vector(user_embedding),
                        Vector(user_embedding),
                        user_id,
                        user_id,
                        job_ids,
                        Vector(user_embedding),
                    ),
                )
                rows: list[dict[str, Any]] = cur.fetchall()

        result: list[tuple[Job, float]] = []
        for row in rows:
            result.append(
                (
                    Job(
                        id=row["id"],
                        source=row["source"] or "kwork",
                        url=row["url"] or "",
                        title=row["title"] or "",
                        description=row["description"],
                        budget=row["budget"] or "",
                        raw_html=row["raw_html"] or "",
                        technologies=list(row["technologies"] or []),
                        posted_at=row["posted_at"],
                        created_at=row["created_at"],
                        match_score=float(row["match_score"]),
                        raw_similarity=float(row["raw_similarity"]),
                        final_score=float(row["final_score"]),
                        ranker_version=row["ranker_version"] or "",
                        reason_codes=list(row["reason_codes"] or []),
                    ),
                    float(row["similarity"]),
                )
            )
        return result
